File: backend/app/schemas/progress.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel

from app.schemas.profile import JLPTLevel

# The full skill list from the master spec's learner model. Only vocabulary
# and grammar have a data source as of Phase 5 — the rest always report
# has_data=False until a later phase adds kanji/reading/listening/speaking/
# conversation content or activities.
TRACKED_CATEGORIES = [
    "vocabulary",
    "grammar",
    "kanji",
    "reading",
    "listening",
    "speaking",
    "conversation",
]


class CategoryProgress(BaseModel):
    category: str
    mastery: float | None
    concepts_tracked: int
    has_data: bool


class OverallProgress(BaseModel):
    mastery: float | None
    has_data: bool


class JLPTReadiness(BaseModel):
    jlpt_target: JLPTLevel | None
    score: float | None
    has_data: bool


class ProgressResponse(BaseModel):
    overall: OverallProgress
    skills: list[CategoryProgress]
    estimated_jlpt_readiness: JLPTReadiness


class MistakeEntry(BaseModel):
    category: str
    concept: str
    occurrences: int
    mastery: float
    last_seen: datetime


def mistake_to_public(doc: dict[str, Any]) -> MistakeEntry:
    correct = doc.get("correct_count", 0)
    incorrect = doc.get("incorrect_count", 0)
    return MistakeEntry(
        category=doc["category"],
        concept=doc["concept"],
        occurrences=incorrect,
        mastery=correct / (correct + incorrect),
        last_seen=doc["last_seen"],
    )

File: frontend/src/types/progress.ts
import type { JlptLevel } from './profile'

export interface CategoryProgress {
  category: string
  mastery: number | null
  concepts_tracked: number
  has_data: boolean
}

export interface OverallProgress {
  mastery: number | null
  has_data: boolean
}

export interface JlptReadiness {
  jlpt_target: JlptLevel | null
  score: number | null
  has_data: boolean
}

export interface ProgressResponse {
  overall: OverallProgress
  skills: CategoryProgress[]
  estimated_jlpt_readiness: JlptReadiness
}

export interface MistakeEntry {
  category: string
  concept: string
  occurrences: number
  mastery: number
  last_seen: string
}

export const CATEGORY_LABELS: Record<string, string> = {
  vocabulary: 'Vocabulary',
  grammar: 'Grammar',
  kanji: 'Kanji',
  reading: 'Reading',
  listening: 'Listening',
  speaking: 'Speaking',
  conversation: 'Conversation',
}
